rcl/src/rcl/init.c

```diff
@@ -162,6 +162,10 @@ rcl_init(
       fail_ret = ret;
       goto fail;
     }
+  } else if (RMW_LOCALHOST_ONLY_DISABLED == *localhost_only) {
+    // If the user sets the rmw init option to RMW_LOCALHOST_ONLY_DISABLED
+    // manually, save the boolean variable as 0
+    *localhost_only = 0;
```

Comment:
why?
`context->impl->init_options.impl->rmw_init_options.localhost_only` is of type `rmw_localhost_only_t`

what needs to be fixed is `rcl_get_localhost_only`

Reply:
`rcl_get_localhost_only` is not called if a user sets `localhost_only=RMW_LOCALHOST_ONLY_DISABLED` in the `rmw_init_options_t` struct. That's the case being covered by this portion of code.

Reply:
Yes, in that case `rcl_get_localhost_only` is not being called, and there's no need to modify `localhost_only` value either (i.e. when `context->impl->init_options.impl->rmw_init_options.localhost_only == RMW_LOCALHOST_ONLY_DISABLED` things are working well).

They are not working well when `context->impl->init_options.impl->rmw_init_options.localhost_only == RMW_LOCALHOST_ONLY_DEFAULT` because of [this line](https://github.com/ros2/rcl/blob/e47042cba4f32f77b361d0f87d905b9386cadde8/rcl/src/rcl/localhost.c#L44). That needs to be corrected.

Reply:
> They are not working well when `context->impl->init_options.impl->rmw_init_options.localhost_only == RMW_LOCALHOST_ONLY_DEFAULT` because of [this line](https://github.com/ros2/rcl/blob/e47042cba4f32f77b361d0f87d905b9386cadde8/rcl/src/rcl/localhost.c#L44). That needs to be corrected.

That line is returning 1 or 0 properly (although its signature is misleading and has to be changed, left an issue opened for that).

But I do think it's not OK that all the functions receiving that enum have to make the boolean comparison against `RMW_LOCALHOST_ONLY_DISABLED` before using the value and IMO this should be addressed in this function where the context is initialized.

I'll remove this because the check you suggested me for `node.c` solves the case I was addressing but it would be great to come back at this at some point.

Reply:
> That line is returning 1 or 0 properly

It's returning `1` or `0` and setting that to a `rmw_localhost_only_t` which doesn't expect `1/0`, so that line is wrong. It should be:

```c
  *localhost_only = (ros_local_host_env_val != NULL && strcmp(ros_local_host_env_val, "1") == 0) ? RMW_LOCALHOST_ONLY_ENABLED ? RMW_LOCALHOST_ONLY_DISABLED;
```

>  (although its signature is misleading and has to be changed, left an issue opened for that).

The signature is ok and shouldn't be changed. The idea is that in the future we will have the following functions:

```c
rcl_init_options_enable_localhost_only(rcl_init_options_t * init_options);
rcl_init_options_disable_localhost_only(rcl_init_options_t * init_options);
```

There would be equivalents in `rclcpp/rclpy` for those.
If the default constructed `init_options` isn't modified with those functions, localhost only will be set according to the environment variable (as by default it's value is RMW_LOCALHOST_ONLY_DEFAULT).
If the user explicitly enabled/disabled it, the environment variable won't be checked.

As a consequence, at `rcl` level the variable can be `default/enable/disable`, and at `rmw` implementation level can be only `enable/disable`.

Reply:
I'll open a PR for that in the `rcl_get_localhost_only` function. This PR is about passing the context setting to the node, mind giving your thoughts on it?

Reply:
> I'll open a PR for that in the `rcl_get_localhost_only` function. This PR is about passing the context setting to the node, mind giving your thoughts on it?

I don't see why opening another PR, as I think this change is incorrect and the only problem is in `rcl_get_localhost_only`.
Can you explain what this extra `else/if` does?

> This PR is about passing the context setting to the node

Can you explain what you mean?

Reply:
> I don't see why opening another PR, as I think this change is incorrect and the only problem is in `rcl_get_localhost_only`.

IMO there are two problems:

1. The one fixed in this PR, `localhost_only` being loaded from environment variable in `node.c` although this variable was already loaded in init.c

2. rcl_get_localhost_only returning boolean instead of  rmw_localhost_only_t.

This PR fixes problem 1. 

> Can you explain what this extra `else/if` does?

Based on my conversation with you, that else/if was removed, please check the latest commit of this PR.

Reply:
> rcl_get_localhost_only returning boolean instead of rmw_localhost_only_t.

It's not returning a boolean, it's returning `rmw_localhost_only_t` and doing the conversion incorrectly.
IMO, we should fix the logic in this same PR as today is the deadline for merging things before Foxy release (anyways we can backport the fix in a patch release, so it's not a big deal).

> Based on my conversation with you, that else/if was removed, please check the latest commit of this PR.

I didn't see the commit that removed that :joy:.